### PR TITLE
`<ArrayInput>` uses `SourceContext` instead of cloning children

### PR DESCRIPTION
**Pull request by adguernier**
## Problem

`<ArrayInput>` and `<SimpleFormIterator>` are still cloning their children. Besides, `<SimpleFormIterator>` throws an Error when used without a source prop.

## Solution

Migrate both components to support the `SourceContext` and no longer require cloning their children.

This is a BC so this needs to be done before the v5 release.

### CODE REVIEW

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx [hunk 1]
```diff
@@ -177,15 +198,13 @@ export const ArrayInput = (props: ArrayInputProps) => {
...
-                    record,
-                    resource: resourceFromProps,
-                    source,
-                    variant,
-                    margin,
-                    disabled,
-                })}
+                <ResourceContextProvider value={resourceFromProps}>
+                    <SourceContextProvider value={sourceContext}>
+                        <RecordContextProvider value={record}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx:203]: Use `OptionalRecordContextProvider` instead

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx [hunk 2]
```diff
@@ -141,6 +147,21 @@ export const ArrayInput = (props: ArrayInputProps) => {
...
 
+    const sourceContext = React.useMemo<SourceContextValue>(
+        () => ({
+            getSource: (source: string) => {
+                return source ? `${finalSource}.${source}` : finalSource;
+            },
+            getLabel: (source: string) =>
+                parentSourceContext
+                    ? parentSourceContext.getLabel(
+                          `resources.${resourceFromProps}.fields.${source}`
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx:158]: This is mysterious to me. Would you mind adding a comment to give an example expected label? 

Because to me, inside another ArrayInput, this will lead to labels like `resources.A.fields.B.0.resources.A.fields.C` instead of `resources.A.fields.B.0.fields.C`

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx [hunk 3]
```diff
@@ -92,6 +97,7 @@ export const ArrayInput = (props: ArrayInputProps) => {
     const formGroupName = useFormGroupContext();
     const formGroups = useFormGroups();
     const finalSource = useWrappedSource(source);
+    const parentSourceContext = useSourceContext();
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx:100]: I find it strange to call both `useWrappedSource` and `useSourceContext` in the same file, as the first one calls the second one.

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx [hunk 4]
```diff
@@ -141,6 +147,21 @@ export const ArrayInput = (props: ArrayInputProps) => {
...
+        () => ({
+            getSource: (source: string) => {
+                return source ? `${finalSource}.${source}` : finalSource;
+            },
+            getLabel: (source: string) =>
+                parentSourceContext
+                    ? parentSourceContext.getLabel(
+                          `resources.${resourceFromProps}.fields.${source}`
+                      )
+                    : `resources.${resourceFromProps}.fields.${source}`,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx:160]: use the `resource` from `ResourceContext` instead of `resourceFromProps`

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx [hunk 5]
```diff
@@ -141,6 +147,21 @@ export const ArrayInput = (props: ArrayInputProps) => {
 
     const { error } = getFieldState(finalSource, formState);
 
+    const sourceContext = React.useMemo<SourceContextValue>(
+        () => ({
+            getSource: (source: string) => {
+                return source ? `${finalSource}.${source}` : finalSource;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx:153]: This I don't understand. In other instances, we check if there is a parent source context, then call that parent's getSource, otherwise return the source. Here you're doing the exact opposite...

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx [hunk 6]
```diff
@@ -165,14 +172,14 @@ export const SimpleFormIterator = (inProps: SimpleFormIteratorProps) => {
...
-                            member={`${source}.${index}`}
+                            member={`${index}`}
                             onRemoveField={removeField}
                             onReorder={handleReorder}
                             record={(records && records[index]) || {}}
                             removeButton={removeButton}
                             reOrderButtons={reOrderButtons}
                             resource={resource}
-                            source={source}
+                            source={finalSource}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx:182]: We should stop passing the source altogether, and let the item read the sourceContext instead

**Review by fzaninotto — CHANGES_REQUESTED** [review]
I'm confused by this PR. It probably requires more tests and more comments to make the code robust and maintainable.

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 7]
```diff
@@ -46,6 +53,14 @@ export const TranslatableFieldsTabContent = (
     );
     const addLabel = Children.count(children) > 1;
 
+    // As fields rely on the RecordContext to get their values and have no knowledge of the locale,
+    // we need to create a new record with the values for the current locale only
+    // Given the record { title: { en: 'title_en', fr: 'title_fr' } } and the locale 'fr',
+    // the record for the locale 'fr' will be { title: 'title_fr' }
+    const [recordForLocale] = useState(() =>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:60]: I don't understand why you use useState here.

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 8]
```diff
@@ -109,3 +119,31 @@ const Root = styled('div', {
...
+            const value = get(record, path);
+            return set(acc, pathWithoutLocale, value);
+        }
+        return acc;
+    }, structuredClone(record));
+
+    return recordForLocale;
+};
+
+const getRecordPaths = (
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:141]: Please add a comment to explain the output for a given input

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 9]
```diff
@@ -30,7 +37,7 @@ export const TranslatableFieldsTabContent = (
     } = props;
     const { selectedLocale } = useTranslatableContext();
 
-    const parentSourceContext = useSourceContext();
+    const parentSourceContext = useOptionalSourceContext();
     const sourceContext = React.useMemo(
         () => ({
             getSource: (source: string) =>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:43]: We shouldn't do anything special here. The sourceContext is only useful for the label.

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 10]
```diff
@@ -109,3 +119,31 @@ const Root = styled('div', {
...
 }));
+
+const getRecordForLocale = (record: RaRecord, locale: string) => {
+    // Get all paths of the record
+    const paths = getRecordPaths(record);
+
+    // For each path, if a path ends with the locale, set the value of the path without the locale
+    // to the value of the path with the locale
+    const recordForLocale = paths.reduce((acc, path) => {
+        if (path.includes(locale)) {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:130]: This is too weak IMO, as it will match 'francois' for the 'fr' locale. I'd use a RegExp, testing for entries like 'XXX.[locale]' only.

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx [hunk 11]
```diff
@@ -56,9 +56,15 @@ export const SimpleFormIterator = (inProps: SimpleFormIteratorProps) => {
         fullWidth,
         sx,
     } = props;
-    if (!source) {
-        throw new Error('SimpleFormIterator requires a source prop');
+
+    const sourceContext = useSourceContext();
+    const finalSource = sourceContext?.getSource('');
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx:61]:
```suggestion
    const finalSource = sourceContext.getSource('');
```

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.spec.tsx [hunk 12]
```diff
@@ -17,36 +17,13 @@ import { ArrayInput } from './ArrayInput';
 import { SimpleFormIterator } from './SimpleFormIterator';
 import { useFormContext } from 'react-hook-form';
 import { GlobalValidation, ValidationInFormTab } from './ArrayInput.stories';
+import { useArrayInput } from './useArrayInput';
 
 describe('<ArrayInput />', () => {
-    it('should pass its record props to its child', async () => {
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.spec.tsx:22]: That's a problem that it doesn't work. Without a RecordContext, we can't use a field as child of ArrayInput.

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx [hunk 13]
```diff
@@ -156,36 +199,32 @@ export const ArrayInput = (props: ArrayInputProps) => {
...
-                `ra-input-${source}`,
+                `ra-input-${finalSource}`,
                 ArrayInputClasses.root,
                 className
             )}
             error={!!error}
             {...sanitizeInputRestProps(rest)}
         >
             <InputLabel
-                htmlFor={finalSource}
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx:167]: why did you remove that?

File: test-setup.js [hunk 14]
```diff
@@ -17,3 +17,8 @@ const { Response, Headers, Request } = require('whatwg-fetch');
 global.Response = Response;
 global.Headers = Headers;
 global.Request = Request;
+
+// jsdom does not support structuredClone
```
Comment (fzaninotto) [inline test-setup.js:21]: All our users will have to do that. Please use a shim instead.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx (hunk 10 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:130]: `path` is an array, not a string

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.spec.tsx (hunk 12 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.spec.tsx:22]: ArrayInput does not set the RecordContext, SimpleFormIteratorItem does

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx (hunk 13 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.tsx:167]: Because you can't label a non focusable element

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 15]
```diff
@@ -57,28 +53,21 @@ export const TranslatableFieldsTabContent = (
         >
             {Children.map(children, field =>
                 field && isValidElement<any>(field) ? (
-                    <ResourceContextProvider
-                        value={resource}
-                        key={field.props.source}
-                    >
-                        <RecordContextProvider value={record}>
-                            <SourceContextProvider value={sourceContext}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:65]: We still need a SourceContextProvider to have the label translated

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 16]
```diff
@@ -28,23 +30,17 @@ export const TranslatableFieldsTabContent = (
...
+    const addLabel = Children.count(children) > 1;
 
-    const parentSourceContext = useSourceContext();
-    const sourceContext = React.useMemo(
-        () => ({
-            getSource: (source: string) =>
-                parentSourceContext
-                    ? parentSourceContext.getSource(`${source}.${locale}`)
-                    : `${source}.${locale}`,
-            getLabel: (source: string) =>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:40]: I think this is still required

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx [hunk 17]
```diff
@@ -56,9 +56,14 @@ export const SimpleFormIterator = (inProps: SimpleFormIteratorProps) => {
         fullWidth,
         sx,
     } = props;
-    if (!source) {
-        throw new Error('SimpleFormIterator requires a source prop');
+
+    const finalSource = useWrappedSource('');
+    if (!finalSource) {
+        throw new Error(
+            'SimpleFormIterator should be wrapped in a SourceContext'
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIterator.tsx:63]:
```suggestion
            'SimpleFormIterator can only be called within an iterator input like ArrayInput'
```

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIteratorItem.tsx [hunk 18]
```diff
@@ -72,27 +71,38 @@ export const SimpleFormIteratorItem = React.forwardRef(
...
+                        // </ArrayInput>
                         return parentSourceContext.getSource(
-                            source ? `${member}.${source}` : member
+                            `${index}.${source}`
                         );
                     }
-                    return source ? `${member}.${source}` : member;
                 },
                 getLabel: (source: string) => {
-                    // remove digits, e.g. 'book.authors.2.categories.3.identifier.name' => 'book.authors.categories.identifier.name'
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIteratorItem.tsx:83]: I think we should keep this comment

**Review by fzaninotto — CHANGES_REQUESTED** [review]
I think we miss a story where we can see non-humanized input labels, to check that they make sense.

File: test-setup.js (hunk 14 above)
Comment (fzaninotto) [reply test-setup.js:21]: We should not use structuredClone.

File: packages/ra-core/src/i18n/useTranslatable.spec.ts [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+                items: [
+                    { description: { en: 'item1_en', fr: 'item1_fr' } },
+                    { description: { en: 'item2_en', fr: 'item2_fr' } },
+                ],
+            };
+
+            const recordForLocale = getRecordForLocale(record, 'fr');
+
+            expect(recordForLocale).toEqual({
+                fractal: true,
```
[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/i18n/useTranslatable.spec.ts:19]: It includes the `fr` letter to make sure we don't mistake it for the `fr` locale. See https://github.com/marmelab/react-admin/pull/9911#discussion_r1634873152

File: packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIteratorItem.tsx (hunk 18 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ArrayInput/SimpleFormIteratorItem.tsx:83]: We have a way better one now and you wrote it ;)

File: packages/ra-core/src/i18n/useTranslatable.spec.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/i18n/useTranslatable.spec.ts:19]: Then your test proves that the algorith is wrong: `fractal` shouldn't be included in the record for the `fr` locale.

[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/i18n/useTranslatable.spec.ts:19]: No, that's actually what I expect, have the record returned with its translatable properties fixed for the locale

[the remark this answers is not in the document]
Comment (djhi) [reply packages/ra-core/src/i18n/useTranslatable.spec.ts:19]: If for some reason, you have a `FormDataConsumer`, you'll still get the full record

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx [hunk 20]
```diff
@@ -392,11 +452,113 @@ export const NestedInline = () => (
...
+                                                sx={{ width: 200 }}
+                                            />
+                                        </Labeled>
+                                    </TranslatableInputs>
+                                    <TranslatableFields locales={['en', 'fr']}>
+                                        <TextField
+                                            source="name"
+                                            sx={{ width: 200 }}
+                                        />
+                                        <TextField
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx:523]: why is it doubled?

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx [hunk 21]
```diff
@@ -392,11 +452,113 @@ export const NestedInline = () => (
...
+                                console.log(data);
+                            },
+                        }}
+                    >
+                        <SimpleForm>
+                            <TextInput source="customer" helperText={false} />
+                            <DateInput source="date" helperText={false} />
+                            <ArrayInput source="items">
+                                <SimpleFormIterator
+                                    inline
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx:503]: Please avoid inline, this makes inputs smaller and truncates the labels

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 22]
```diff
@@ -57,28 +77,25 @@ export const TranslatableFieldsTabContent = (
...
-                                            {field}
-                                        </Labeled>
-                                    ) : (
-                                        field
-                                    )}
-                                </div>
-                            </SourceContextProvider>
-                        </RecordContextProvider>
-                    </ResourceContextProvider>
+                    <RecordContextProvider value={recordForLocale}>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:80]: missing key. Also, it's the same record for all the fields, why do we create the RecordContext inside the Children.map? It should be outside.

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx [hunk 23]
```diff
@@ -24,14 +30,21 @@ export const TranslatableFieldsTabContent = (
...
         ...other
     } = props;
-    const { selectedLocale } = useTranslatableContext();
+    const { selectedLocale, getRecordForLocale } = useTranslatableContext();
+    const addLabel = Children.count(children) > 1;
 
-    const parentSourceContext = useSourceContext();
-    const sourceContext = React.useMemo(
+    const parentSourceContext = useOptionalSourceContext();
+    const resource = useResourceContext(props);
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:41]: unused?

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Getting close! 🔥

File: packages/ra-core/src/i18n/useTranslatable.spec.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/ra-core/src/i18n/useTranslatable.spec.ts:19]: OK, gotcha

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx (hunk 20 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx:523]: Because when there's only one child, the `TranslatableFields` does not add labels

File: packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx (hunk 23 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/field/TranslatableFieldsTabContent.tsx:41]: Nope, it's used in `getLabel`

File: packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx (hunk 20 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/input/ArrayInput/ArrayInput.stories.tsx:523]: A comment in the code will make future you aware of that peculiarity

**Review by fzaninotto — APPROVED** [review]
